src/averageAges.js

```diff
@@ -55,7 +68,37 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const peopleNameList = people.map(person => person.name);
+  let childrenWithMothers
+    = people.filter(person => peopleNameList.includes(person.mother));
+
+  if (onlyWithSon) {
+    childrenWithMothers
+      = childrenWithMothers.filter(person => person.sex === 'm');
+  }
+
+  const averageAgeDiff = calculateAverageAge(childrenWithMothers, people);
+
+  return averageAgeDiff;
+}
+
+function calculateAverageAge(sampleOfPeople, people) {
+  if (!sampleOfPeople.length) {
+    return 0;
+  }
+
+  const ageDifference = people
+    ? sampleOfPeople.map(child => {
+      const mother = people.find(person => person.name === child.mother);
+
+      return child.born - mother.born;
+    })
+    : sampleOfPeople.map(person => person.died - person.born);
+
+  const sumOfAgeDiff = ageDifference.reduce((sum, ageDiff) => sum + ageDiff, 0);
+  const averageAge = sumOfAgeDiff / sampleOfPeople.length;
+
+  return averageAge;
 }
```

Comment:
this function should do only one action - calculate average age, but not define mother for child, because it becomes tricky. take only `ageDifference` as param here and do mapping 
``` 
sampleOfPeople.map(child => {
      const mother = people.find(person => person.name === child.mother);

      return child.born - mother.born;
    })
``` 
there `calculateAverageAgeDiff`